File: apps/api/src/risk/engine.ts
/**
 * Risk Engine — TypeScript port of pipeline_py/risk/engine.py.
 *
 * The Python module is the batch-side reference (its test suite is the
 * source of truth for the math); this port exists because Cloudflare
 * Workers can't run Python and the Worker is where a user's live decision
 * happens — portfolio heat and open-position counts change the moment a
 * position opens or closes, so a once-daily precomputed verdict would be
 * stale within the same session. Keep the two in lockstep: any gate added
 * here needs the matching gate in engine.py, and vice versa.
 *
 * Every gate is arithmetic — R:R (net of execution cost), lot size,
 * notional, portfolio heat, open-position count, sector concentration,
 * turnover, liquidity. None of it depends on a Setup or Regime being
 * statistically valid; unproven ideas get recorded, not gated on.
 */
import { planEconomics, type PlanEconomics } from './cost';

export const LOT_SIZE = 100;
export const MIN_TURNOVER = 300_000_000; // matches pipeline_py/screening/scoring.py

export const PASS_GATES = ['min_rr', 'lot_size', 'notional', 'turnover', 'liquidity'] as const;
export const WAIT_GATES = ['heat', 'max_positions', 'sector_concentration'] as const;

export interface Account {
  capital: number;
  risk_per_trade: number;
  max_positions: number;
  max_heat: number;
  max_notional_pct: number;
  min_rr: number;
  max_sector_positions: number;
}

export interface PortfolioContext {
  openPositions: number;
  currentHeat: number;
  sectorPositionCounts: Record<string, number>;
}

export interface PlanLevels {
  triggerPrice: number;
  stopPlanned: number;
  targetPlanned: number;
  sector33?: string | null;
  turnoverValue?: number | null;
  avgVolume20d?: number | null;
  /** Drives the cost model. Absent, cost falls back to the flat slippage
   * floor, which understates it on volatile names — the resulting R:R is
   * then optimistic. Callers that can supply ATR should. */
  atr?: number | null;
}

export interface GateResult {
  passed: boolean;
  value: number | null;
  threshold?: number;
  /** min_rr only: the pre-cost figure, kept beside the net one so the gap
   * is visible rather than folded away. */
  gross?: number;
}

export type Decision = 'BUY' | 'WAIT' | 'PASS';

export interface RiskVerdict {
  decision: Decision;
  reasonCode: string;
  shares: number;
  riskAmount: number;
  riskPct: number;
  notional: number;
  rrGross: number;
  rrNet: number;
  gates: Record<string, GateResult>;
  economics: PlanEconomics | null;
}

/** Shares sized off the stop distance, floored to the lot size. Flooring
 * only ever reduces the risk below the target — never rounds up past it —
 * so risk_amount is a ceiling, not an approximation. */
export function positionSize(
  capital: number,
  riskPerTrade: number,
  trigger: number,
  stop: number
): { shares: number; riskAmount: number } {
  const riskPerShare = trigger - stop;
  if (riskPerShare <= 0) return { shares: 0, riskAmount: 0 };
  const targetRisk = capital * riskPerTrade;
  const rawShares = targetRisk / riskPerShare;
  const shares = Math.floor(rawShares / LOT_SIZE) * LOT_SIZE;
  return { shares, riskAmount: shares * riskPerShare };
}

export function evaluateRisk(
  account: Account,
  levels: PlanLevels,
  ctx: PortfolioContext
): RiskVerdict {
  const gates: Record<string, GateResult> = {};

  // The R:R gate is evaluated NET of execution cost. Gating on the gross
  // figure approves plans whose advertised edge is entirely consumed by
  // getting in and out — on this repo's own data the round trip runs
  // 0.10-0.18R, larger than any selection effect measured here. Cost is
  // arithmetic, not a hypothesis, so it is allowed to block.
  const econ = planEconomics(
    levels.triggerPrice,
    levels.stopPlanned,
    levels.targetPlanned,
    levels.atr
  );
  const rrGross = econ ? econ.rrGross : 0;
  const rrNet = econ ? econ.rrNet : 0;
  gates.min_rr = {
    passed: rrNet >= account.min_rr,
    value: round(rrNet, 3),
    gross: round(rrGross, 3),
    threshold: account.min_rr,
  };

  const { shares, riskAmount } = positionSize(
    account.capital,
    account.risk_per_trade,
    levels.triggerPrice,
    levels.stopPlanned
  );
  gates.lot_size = { passed: shares >= LOT_SIZE, value: shares };

  const notional = shares * levels.triggerPrice;
  const maxNotional = account.capital * account.max_notional_pct;
  gates.notional = {
    passed: notional <= maxNotional,
    value: round(notional, 2),
    threshold: round(maxNotional, 2),
  };

  const riskPct = account.capital ? riskAmount / account.capital : 0;
  const newHeat = ctx.currentHeat + riskPct;
  gates.heat = { passed: newHeat <= account.max_heat, value: round(newHeat, 4), threshold: account.max_heat };

  gates.max_positions = {
    passed: ctx.openPositions < account.max_positions,
    value: ctx.openPositions,
    threshold: account.max_positions,
  };

  const sectorCount = levels.sector33 ? ctx.sectorPositionCounts[levels.sector33] ?? 0 : 0;
  gates.sector_concentration = {
    passed: sectorCount < account.max_sector_positions,
    value: sectorCount,
    threshold: account.max_sector_positions,
  };

  // Unknown turnover/volume doesn't block — matches score_theme's own
  // _passes_liquidity, which treats missing data as "not disqualifying".
  if (levels.turnoverValue != null) {
    gates.turnover = {
      passed: levels.turnoverValue >= MIN_TURNOVER,
      value: levels.turnoverValue,
      threshold: MIN_TURNOVER,
    };
  } else {
    gates.turnover = { passed: true, value: null };
  }

  if (levels.avgVolume20d && shares) {
    const advPct = shares / levels.avgVolume20d;
    gates.liquidity = { passed: advPct <= 0.01, value: round(advPct, 4) };
  } else {
    gates.liquidity = { passed: true, value: null };
  }

  const verdict = (decision: Decision, reasonCode: string): RiskVerdict => ({
    decision,
    reasonCode,
    shares,
    riskAmount,
    riskPct,
    notional,
    rrGross,
    rrNet,
    gates,
    economics: econ,
  });

  for (const gate of PASS_GATES) {
    if (!gates[gate].passed) {
      // A plan whose gross R:R cleared the bar and whose net one did not
      // failed for a reason the user can act on — widen the stop, or accept
      // that this trade is too tight to be worth taking — so it gets its own
      // reason rather than reading as an ordinary bad-R:R rejection.
      if (gate === 'min_rr' && rrGross >= account.min_rr) {
        return verdict('PASS', 'min_rr_after_cost');
      }
      return verdict('PASS', gate);
    }
  }
  for (const gate of WAIT_GATES) {
    if (!gates[gate].passed) return verdict('WAIT', gate);
  }

  return verdict('BUY', 'all_gates_passed');
}

function round(v: number, digits: number): number {
  const m = 10 ** digits;
  return Math.round(v * m) / m;
}

File: pipeline_py/risk/engine.py
"""Risk Engine: turns a trade plan's price levels into a position size and a
BUY / WAIT / PASS verdict.

Every gate here is arithmetic: R:R (net of execution cost), lot size,
notional, portfolio heat, open-position count, sector concentration,
turnover, liquidity. None of it
requires an unproven Setup or Regime to be right — that is deliberate. The
design's central rule is that unproven ideas get recorded, not gated on;
only calculations that don't need statistical validation are allowed to
block a trade. If a gate here ever depends on "is this Setup any good",
it belongs in analytics, not in this module.

PASS vs WAIT is a real distinction, not two names for "no": PASS is about
the instrument (this name doesn't qualify on its own terms — bad R:R, too
illiquid, too small to fill at the lot size). WAIT is about the portfolio
(the name would be fine on its own, but the book has no room — heat, slot
count, sector concentration). A WAITed plan can still fire later if the
book frees up; a PASSed one should not resurface untouched.
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional

from ..screening.scoring import MIN_TURNOVER
from .cost_in_r import PlanEconomics, plan_economics

LOT_SIZE = 100

# Ordered so the first failing gate is the one reported: instrument-level
# reasons take priority over portfolio-level ones, since "this name is bad"
# is a more useful thing to tell the user than "no room right now" when both
# happen to be true simultaneously.
PASS_GATES = ("min_rr", "lot_size", "notional", "turnover", "liquidity")
WAIT_GATES = ("heat", "max_positions", "sector_concentration")


@dataclass(frozen=True)
class PortfolioContext:
    """What is already true about the book before this plan is considered.
    Assembled by the caller from `positions` — this module has no DB access,
    so the same verdict logic is exercised the same way in tests and prod."""

    open_positions: int
    current_heat: float  # fraction of capital already at risk
    sector_position_counts: dict[str, int]


@dataclass(frozen=True)
class PlanLevels:
    trigger_price: float
    stop_planned: float
    target_planned: float
    sector33: Optional[str] = None
    turnover_value: Optional[float] = None
    avg_volume_20d: Optional[float] = None
    # Drives the cost model. Absent, cost falls back to the flat slippage
    # floor, which understates it on volatile names — so the resulting R:R
    # is optimistic rather than wrong in the safe direction. Callers that
    # can supply ATR should.
    atr: Optional[float] = None


@dataclass(frozen=True)
class RiskVerdict:
    decision: str  # BUY | WAIT | PASS
    reason_code: str
    shares: int
    risk_amount: float
    risk_pct: float
    notional: float
    rr_gross: float
    rr_net: float
    gates: dict[str, dict]
    economics: Optional[PlanEconomics] = None


def position_size(
    capital: float, risk_per_trade: float, trigger: float, stop: float
) -> tuple[int, float]:
    """Shares sized off the stop distance, floored to the lot size.

    Returns (shares, actual_risk_amount). Flooring to the lot only ever
    reduces the risk below the target — it never rounds up past the cap —
    so the caller can trust risk_amount is a ceiling, not an approximation
    that might run hot.
    """
    risk_per_share = trigger - stop
    if risk_per_share <= 0:
        return 0, 0.0
    target_risk = capital * risk_per_trade
    raw_shares = target_risk / risk_per_share
    shares = int(raw_shares // LOT_SIZE) * LOT_SIZE
    return shares, shares * risk_per_share


def evaluate(account: dict, levels: PlanLevels, ctx: PortfolioContext) -> RiskVerdict:
    """account: a row from `accounts` (capital, risk_per_trade, max_positions,
    max_heat, max_notional_pct, min_rr, max_sector_positions)."""
    gates: dict[str, dict] = {}

    # The R:R gate is evaluated NET of execution cost. Gating on the gross
    # figure would approve plans whose advertised edge is entirely consumed
    # by getting in and out — on this repo's own data the round trip runs
    # 0.10-0.18R, which is larger than any selection effect measured here.
    # Cost is arithmetic, not a hypothesis, so it is allowed to block.
    econ = plan_economics(
        levels.trigger_price, levels.stop_planned, levels.target_planned, levels.atr
    )
    rr_gross = econ.rr_gross if econ else 0.0
    rr_net = econ.rr_net if econ else 0.0
    gates["min_rr"] = {
        "passed": rr_net >= account["min_rr"],
        "value": round(rr_net, 3),
        "gross": round(rr_gross, 3),
        "threshold": account["min_rr"],
    }

    shares, risk_amount = position_size(
        account["capital"], account["risk_per_trade"], levels.trigger_price, levels.stop_planned
    )
    gates["lot_size"] = {"passed": shares >= LOT_SIZE, "value": shares}

    notional = shares * levels.trigger_price
    max_notional = account["capital"] * account["max_notional_pct"]
    gates["notional"] = {
        "passed": notional <= max_notional,
        "value": round(notional, 2),
        "threshold": round(max_notional, 2),
    }

    risk_pct = risk_amount / account["capital"] if account["capital"] else 0.0
    new_heat = ctx.current_heat + risk_pct
    gates["heat"] = {
        "passed": new_heat <= account["max_heat"],
        "value": round(new_heat, 4),
        "threshold": account["max_heat"],
    }

    gates["max_positions"] = {
        "passed": ctx.open_positions < account["max_positions"],
        "value": ctx.open_positions,
        "threshold": account["max_positions"],
    }

    sector_count = (
        ctx.sector_position_counts.get(levels.sector33, 0) if levels.sector33 else 0
    )
    gates["sector_concentration"] = {
        "passed": sector_count < account["max_sector_positions"],
        "value": sector_count,
        "threshold": account["max_sector_positions"],
    }

    # Unknown turnover/volume doesn't block — matches the screen's own
    # _passes_liquidity, which treats missing data as "not disqualifying"
    # rather than as a failure. A stricter reading would double-penalize
    # names the ingest pipeline just hasn't backfilled turnover for yet.
    if levels.turnover_value is not None:
        gates["turnover"] = {
            "passed": levels.turnover_value >= MIN_TURNOVER,
            "value": levels.turnover_value,
            "threshold": MIN_TURNOVER,
        }
    else:
        gates["turnover"] = {"passed": True, "value": None}

    if levels.avg_volume_20d and shares:
        adv_pct = shares / levels.avg_volume_20d
        gates["liquidity"] = {"passed": adv_pct <= 0.01, "value": round(adv_pct, 4)}
    else:
        gates["liquidity"] = {"passed": True, "value": None}

    def verdict(decision: str, reason: str) -> RiskVerdict:
        return RiskVerdict(
            decision, reason, shares, risk_amount, risk_pct, notional, rr_gross, rr_net, gates, econ
        )

    for gate in PASS_GATES:
        if not gates[gate]["passed"]:
            # A plan whose gross R:R cleared the bar and whose net one did not
            # failed for a reason the user can act on — widen the stop, or
            # accept that this particular trade is too tight to be worth
            # taking — so it gets its own reason rather than reading as an
            # ordinary bad-R:R rejection.
            if gate == "min_rr" and rr_gross >= account["min_rr"]:
                return verdict("PASS", "min_rr_after_cost")
            return verdict("PASS", gate)
    for gate in WAIT_GATES:
        if not gates[gate]["passed"]:
            return verdict("WAIT", gate)

    return verdict("BUY", "all_gates_passed")
